### ISSUE TITLE
MQTTv3: connect hangs indefinitely if broker is not reachable

### ISSUE THREAD

**Issue by mcollovati** [eclipse-paho/paho.mqtt.java#843]
Please fill out the form below before submitting, thank you!

- [x] Bug exists Release Version 1.2.5 ( Master Branch)
- [ ] Bug exists in MQTTv3 Client on Snapshot Version 1.2.6-SNAPSHOT (Develop Branch)
- [ ] Bug exists in MQTTv5 Client on Snapshot Version 1.2.6-SNAPSHOT (Develop Branch)

In paho 1.2.5 the call to `MqttClient.connect` hangs indefinitely, regardless the connection timeout settings,
if the broker is not reachable.

In 1.2.4 the method failed with `java.io.IOException: Connection is lost`.

The problem seems to be related to issue #719.

```
// paho 1.2.4
if (message != null) {
	// A new message has arrived
	clientState.notifyReceivedMsg(message);
}  
else {
    // fix for bug 719
	if (!clientComms.isConnected()) {
		throw new IOException("Connection is lost.");
	}
}


// paho 1.2.5
if (message != null) {
	// A new message has arrived
	clientState.notifyReceivedMsg(message);
}  
else {
	// fix for bug 719
	if (!clientComms.isConnected() && !clientComms.isConnecting()) {
        throw new IOException("Connection is lost.");
	}
}
```

During connection attempt `clientComms` is in status `connecting`; 
when broker is down a null message is received and, in paho 1.2.4, an exception is immediately thrown because `clientComms` is not connected.
In 1.2.5 however the exception is never throw because `clientComms` is in connecting state and this leads to an infinite loop.

I don't know if this is a bug or if the problem can be avoided in some way.

**Comment by sp193** [comment]
Have you seen #763 (which succeeded #719)? At the very end of it, it looks like it solved the original issue for #719.

As for where this code change came from, it is not clear to me. It looks like it got copied from the MQTTv5 client (#797), which was added as part of a large patch (#795). Tagging @rdasgupt in case he has any comment about this.

**Comment by mcollovati** [comment]
Thanks for your answer @sp193.
I've seen #763, but it seems to me this situation is a little bit different,but, please, correct me if I'm wrong.
In the mentioned issue at the end the broker is up and can be contacted; in my case the broker is completely down and there will be no way to establish a connection.
I think it is more similar to #778, but here connection hangs forever regardless the socket timeout settings, because the broker is down and there is no way to exit the **CommsReceiver.run** loop.

**Comment by sp193** [comment]
What I meant to say is that if you now experience an issue due to the fix introduced with #797, then perhaps the fix was correct to address an issue like the one described in #763, but has caused another bug.

**Comment by mcollovati** [comment]
Just discovered that it was an error on my test.
I'm using ActiveMQ and I set up a **BrokerService** with an mqtt connector and I did not start the **BrokerService** to simulate broker down. 

```java
        broker = new BrokerService();
        broker.setPersistent(false);
        broker.addConnector("mqtt://localhost:" + mqttPort);
        // broker.start() not called
```

However when adding the connector the server socket is immediately created, even if BorkerService is not started; so paho correctly connected to the socket but did not receive any message, hanging up in **CommsReceiver.run** loop .

Sorry for wasting your time
